src/services/services.ts

```diff
@@ -6259,6 +6373,187 @@ namespace ts {
+                        getReferenceEntryForShorthandPropertyAssignment(refNode, typeChecker, result);
+                    }
+
+                    // Check if the node is within an extends or implements clause
+                    const containingHeritageClause = getContainingClassHeritageClause(refNode);
+                    if (containingHeritageClause) {
+                        result.push(getReferenceEntryFromNode(containingHeritageClause.parent));
+                        return;
+                    }
+
+                    // If we got a type reference, try and see if the reference applies to any expressions that can implement an interface
+                    const containingTypeReference = getContainingTypeReference(refNode);
+                    if (containingTypeReference) {
+                        const parent = containingTypeReference.parent;
+                        if (isVariableLike(parent) && parent.type === containingTypeReference && parent.initializer && isImplementationExpression(parent.initializer)) {
+                            maybeAdd(getReferenceEntryFromNode(parent.initializer));
+                        }
+                        else if (isFunctionLike(parent) && parent.type === containingTypeReference && parent.body && parent.body.kind === SyntaxKind.Block) {
+                            forEachReturnStatement(<Block>parent.body, (returnStatement) => {
+                                if (returnStatement.expression && isImplementationExpression(returnStatement.expression)) {
+                                    maybeAdd(getReferenceEntryFromNode(returnStatement.expression));
+                                }
+                            });
+                        }
+                        else if (isTypeAssertionExpression(parent) && isImplementationExpression(parent.expression)) {
+                            maybeAdd(getReferenceEntryFromNode(parent.expression));
+                        }
+                    }
+                }
+
+                // Type nodes can contain multiple references to the same type. For example:
+                //      let x: Foo & (Foo & Bar) = ...
+                // Because we are returning the implementation locations and not the identifier locations,
+                // duplicate entries would be returned here as each of the type references is part of
+                // the same implementation. For that reason, check before we add a new entry
+                function maybeAdd(a: ReferenceEntry) {
+                    if (!forEach(result, b => a.fileName === b.fileName && a.textSpan.start === b.textSpan.start && a.textSpan.length === b.textSpan.length)) {
+                        result.push(a);
+                    }
+                }
+            }
+
+            function getSymbolsForComponentTypes(type: UnionOrIntersectionType, result: Symbol[] = []): Symbol[] {

```

Comment:
Document or rename to `getSymbolsForClassAndInterfaceComponents` to indicate that this throws out primitive components and recurses through union/interface components.